connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/expressions/LiteralProtoConverter.scala

```diff
@@ -0,0 +1,157 @@
+      case v: Date => builder.setDate(DateTimeUtils.fromJavaDate(v))
+      case v: Duration => builder.setDayTimeInterval(IntervalUtils.durationToMicros(v))
+      case v: Period => builder.setYearMonthInterval(IntervalUtils.periodToMonths(v))
+      case v: Array[_] => builder.setArray(arrayBuilder(v))
+      case v: CalendarInterval =>
+        builder.setCalendarInterval(calendarIntervalBuilder(v.months, v.days, v.microseconds))
+      case null => builder.setNull(nullType)
+      case _ => unsupported(s"literal $literal not supported (yet).")
+    }
+  }
+
+  /**
+   * Transforms literal value to the `proto.Expression.Literal`.
+   *
+   * @return
+   *   proto.Expression.Literal
+   */
+  def toLiteralProto(literal: Any): proto.Expression.Literal =
+    toLiteralProtoBuilder(literal).build()
+
+  private def componentTypeToProto(clz: Class[_]): proto.DataType = clz match {
+    // primitive types
+    case JShort.TYPE => toConnectProtoType(ShortType)
+    case JInteger.TYPE => toConnectProtoType(IntegerType)
+    case JLong.TYPE => toConnectProtoType(LongType)
+    case JDouble.TYPE => toConnectProtoType(DoubleType)
+    case JByte.TYPE => toConnectProtoType(ByteType)
+    case JFloat.TYPE => toConnectProtoType(FloatType)
+    case JBoolean.TYPE => toConnectProtoType(BooleanType)
+    case JChar.TYPE => toConnectProtoType(StringType)
+
+    // java classes
+    case _ if clz == classOf[LocalDate] || clz == classOf[Date] =>
+      toConnectProtoType(DateType)
+    case _ if clz == classOf[Instant] || clz == classOf[Timestamp] =>
+      toConnectProtoType(TimestampType)
+    case _ if clz == classOf[LocalDateTime] => toConnectProtoType(TimestampNTZType)
+    case _ if clz == classOf[Duration] => toConnectProtoType(DayTimeIntervalType.DEFAULT)
+    case _ if clz == classOf[Period] => toConnectProtoType(YearMonthIntervalType.DEFAULT)
+    case _ if clz == classOf[JBigDecimal] => toConnectProtoType(DecimalType.SYSTEM_DEFAULT)
+    case _ if clz == classOf[Array[Byte]] => toConnectProtoType(BinaryType)
+    case _ if clz == classOf[Array[Char]] => toConnectProtoType(StringType)
+    case _ if clz == classOf[JShort] => toConnectProtoType(ShortType)
+    case _ if clz == classOf[JInteger] => toConnectProtoType(IntegerType)
+    case _ if clz == classOf[JLong] => toConnectProtoType(LongType)
+    case _ if clz == classOf[JDouble] => toConnectProtoType(DoubleType)
+    case _ if clz == classOf[JByte] => toConnectProtoType(ByteType)
+    case _ if clz == classOf[JFloat] => toConnectProtoType(FloatType)
+    case _ if clz == classOf[JBoolean] => toConnectProtoType(BooleanType)
+
+    // other scala classes
+    case _ if clz == classOf[String] => toConnectProtoType(StringType)
+    case _ if clz == classOf[BigInt] || clz == classOf[BigDecimal] =>
+      toConnectProtoType(DecimalType.SYSTEM_DEFAULT)
+    case _ if clz == classOf[CalendarInterval] => toConnectProtoType(CalendarIntervalType)
```

Comment:
Should not be possible to call `toConnectProtoType` directly because the `elementType` of `ArrayType` cannot be determined there